docs/tutorials/auction.md

```diff
@@ -0,0 +1,244 @@
+
+```ts
+constructor(auctioneer: PubKey, auctionDeadline: bigint) {
+    super(...arguments)
+    // the initial bidder is the auctioneer himeself
+    this.bidder = hash160(auctioneer)
+    this.auctioneer = auctioneer
+    this.auctionDeadline = auctionDeadline
+}
+```
+
+When deploying the contract, the auctioneer locked the minimal bid into the contract, and at this time, the highest bidder would be himself.
+
+```ts
+const auction = new Auction(publicKeyAuctioneer, auctionDeadline)
+const deployTx = await auction.deploy(minBid)
+```
+
+## Public Methods
+
+### Bid
+
+In method `public bid(bidder: PubKeyHash, bid: bigint)`, we need to check if the bidder has a higher bid than the previous one. If so, we update the highest bidder in the contract state and refund the previous bidder.
+
+We can read the previous highest bid from the balance of the contract UTXO.
+
+```ts
+const highestBid: bigint = this.ctx.utxo.value
+```
+
+Then it's easy to demand a higher bid.
+
+```ts
+assert(bid > highestBid, 'the auction bid is lower than the current highest bid')
+```
+
+The spending/redeeming tx has these outputs.
+
+- Contract's new state output: records the new bidder and locks the new bid into contract UTXO.
+
+```ts
+// Log the previous highest bidder
+const highestBidder: PubKeyHash = this.bidder
+// Change the address of the highest bidder.
+this.bidder = bidder
+
+// Auction continues with a higher bidder.
+const auctionOutput: ByteString = this.buildStateOutput(bid)
+```
+
+- A refund P2PKH output: pay back the previous bidder.
+
+```ts
+// Refund previous highest bidder.
+const refundOutput: ByteString = Utils.buildPublicKeyHashOutput(highestBidder, highestBid)
+```
+
+- An optional change P2PKH output.
+
+```ts
+let outputs: ByteString = auctionOutput + refundOutput
+// Add change output.
+if (this.changeAmount > 0) {
+    outputs += this.buildChangeOutput()
+}
+```
+
+At last, we require the transaction to have these outputs using `ScriptContext`.
+
+```ts
+assert(hash256(outputs) == this.ctx.hashOutputs, 'hashOutputs check failed')
+```
+
+As `bid` is called continuously, the state of the contract is constantly updated. The highest bidder, and the highest bid as well, is recorded in the latest contract UTXO until the auctioneer closes the auction.
+
+### Close
+
+Method `public close(sig: Sig)` is simple, we require:
+
+- It can only be called by the auctioneer. That is why we need to pass in the caller's signature.
+
+```ts
+// Check signature of the auctioneer.
+assert(this.checkSig(sig, this.auctioneer), 'signature check failed')
+```
+
+- Now the auction deadline has passed
+
+```ts
+assert(this.ctx.locktime >= this.auctionDeadline, 'auction is not over yet')
+```
+
+:::note
+We don't place any constraint on transaction outputs here, because the auctioneer can send the highest bid to any address he controls, which is what we actually want.
+:::
+
```

Comment:
Move the on-chain contract code from Conclusion to here.

Add a section on customizing bid tx between here and Conclusion, and move bidTxBuilder in it.